Migration guide post-merge review (#1000)
data-apis/array-api: spec/draft/migration_guide.md

# Migration Guide

This page is meant to help migrate your codebase to an array API standard

> compliant implementation. The guide is divided into three parts.

Reviewer: 'Compliance' seems more relevant for producers. I guess for consumers it is rather about becoming interoperable with compliant implementations?

Author: added the consumers' perspective as well in the intro sentence.

The first part gives an overview of the {ref}`ecosystem` libraries, that
are helpful in different contexts when working with the array API standard.

> The first part is dedicated for {ref}`array-producers`. If your library

Reviewer: second part

Author: done

> The first part is dedicated for {ref}`array-producers`. If your library

Reviewer: ```suggestion
The first part is dedicated to {ref}`array-producers`. If your library
```

Author: done

mimics, for example, NumPy's or PyTorch's functionality, then you can find in

> the first part additional instructions and guidance on how to ensure

Reviewer: ```suggestion
mimics, for example, NumPy's or PyTorch's functionality, then you can find in
here additional instructions and guidance on how to ensure
```

Author: done

downstream users can easily pick your solution as an array provider for
their system/algorithm.

> The second part delves into details for array API standard compatibility for

Reviewer: third part

Author: done

{ref}`array-consumers`. This pertains to any software that performs
multidimensional array manipulation in Python, such as may be found in
scikit-learn, SciPy, or statsmodels. If your software relies on a certain
array producing library, such as NumPy or JAX, then you can use the second
part to learn how to make it library agnostic and, as a result, interchange

> array namespaces with significantly less friction.

Reviewer: I think I like this wording better, WDYT?
```suggestion
part to learn how to make it library agnostic and, as a result, use
array namespaces interchangeably with significantly less friction.
```

Author: done

(ecosystem)=

## Ecosystem

Apart from the documented standard, the array API ecosystem also provides
a set of tools and packages to help you with the migration process:

(array-api-compat)=

### array-api-compat

GitHub: [array-api-compat](https://github.com/data-apis/array-api-compat)

User group: Array Consumers

Although NumPy, Dask, CuPy, and PyTorch support the array API standard, there

> are still some corner cases where their behavior diverges from the standard.

Reviewer: still not sure if we should claim support in this first phrase, particularly for Dask and PyTorch

Author: right, I removed Dask and PyTorch from here.

`array-api-compat` provides a compatibility layer to cover an additional subset

> of these corner cases. This is also accompanied by a few utility functions fo

Reviewer: ```suggestion
of these corner cases. This is also accompanied by a few utility functions for
```

Author: done

easier introspection into array objects. As an array consumer, you can still

> rely on the original API while having access to the standard compatible one.

Reviewer: ```suggestion
easier introspection into array objects. As an array consumer, you can consume
standard-compliant namespaces as well as the wrapped namespaces in
`array-api-compat` at the same time.
```

Author: done

(array-api-strict)=

### array-api-strict

GitHub: [array-api-strict](https://github.com/data-apis/array-api-strict)

User group: Array Consumers

`array-api-strict` is a library that provides a strict and minimal
implementation of the array API standard. As a consumer, you can use
`array-api-strict` for parametrising tests with it as an array namespace

> to ensure your code uses APIs compliant with the standard.

Reviewer: ```suggestion
implementation of the array API standard. As a consumer, you can use
`array-api-strict` in parametrising tests over the array namespace
to ensure your code uses only APIs which are in the standard.
```

Author: done

```

The first step should be as simple as assigning the `np` namespace to a dedicated
namespace variable. The convention used in the ecosystem is to name it `xp`.
Then, it is vital to ensure that each method and function call is something that

> the array API standard supports. For example, `dot` is present in the NumPy's

Reviewer: ```suggestion
the array API standard supports. For example, `dot` is present in the NumPy
```

Author: done
